CONTRUBUTING.md updates

frli8848 opened the pull request:
I have done some updates of the CONTRUBTION.md file. I have added more details on the workflow, and added sections on how to submit test images and report bugs.

This is more a request for comments than a PR. 

/Fredrik

D4N commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,121 @@
+        $ git pull --rebase exiv2_upstream master
 
-## Pull Request guidelines
+7. and if you are asked to add a patch:
```
This happens relatively rarely, so I don't think it is necessary to add it here.

piponazo replied:
I agree with @D4N. And, again, thinking about the scope of the document, I am not sure if it is a good idea to add details about general git usage or code contribution guidelines here.

D4N commented:
Thanks for the update, this looks very good.

Oddly enough, the build on travis is failing. I have glanced over the logs and it looks like the mac build fails because zlib is not installed correctly, so that is unrelated to your PR. Looks like a conan issue.

clanmills reviewed (commented):
This is a beautiful job.  Thank You very much, Fredrik.  The TOC gives the document Structure and Style.  Thank You.

Comments:
1) We only require the -fsanitize=address stuff as a check before submitting the PR.  It's not always needed.  What you've said is correct.  I'll leave you to decide if you should spell out the steps clone/build/develop/test/commit and only put the -fsanitize=address into test.

2) Should lines 19+20 be a single line `... such as, Wikis ...` And 72+73 `for non-programmers`.  And 137+138 `... bug report then create ...`

3) Can we put the stuff about patches into  Note: below the General Workflow

4) Capitalisation error on `7. And if you are asked...`

5) I don't understand git/github.  I think there's something missing concerning updating your fork if you want to make future submissions.  I believe your fork has to be updated.

6) Responding to requests for changes to the PR.  If you've deleted your branch, you might have "cut yourself off" from making changes.  I'm not sure - something to have clarified.

Thank You very much for this great work.

D4N commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,121 @@
+1. Then start a terminal (or use your favorite git GUI app) and clone your copy of the Exiv2 repo:
+
+        $  git clone https://github.com:your-username/exiv2.git
+
```
Could you mention here that you should create a new branch for your work based on the exiv2 upstream master branch? Something like this:
```
$ git clone https://github.com:your-username/exiv2.git
$ git remote add exiv2_upstream https://github.com/Exiv2/exiv2.git
$ git checkout -b my_feature_branch exiv2_upstream/master --no-track
```
and then work in the `my_feature_branch`.

piponazo replied:
I am not sure if we should add so much details in this document about how to use git. This is something that from my point of view is out of the scope of the document.

frli8848 commented:
@D4N What is happening when you do the: 

$ git checkout -b my_feature_branch exiv2_upstream/master --no-track

The original CONTRIBUTING.md said that: " We prefer rebasing in top of master, instead of merging master into branches." Is this in conflict with the command above or is it a different thing?

clanmills commented:
@Kicer86 Can you review this, please.  This is an important document to establish the exact process for creating a PR.  There are other interesting/useful matters covered - however it's very important that the process to submit a PR is clarified and simple to follow.

@frli8848 I think we should also document a procedure to enable folks to submit a patch.  For straightforward issues (such as your 3 one-line lens changes in February #227), a patch and test image should be sufficient.  We shouldn't require people to know and understand git/GitHub to make a one-line code change.

@frli8848 Can you clarify the following please?  What is exiv2_upstream?
```
After the PR has been reviewed by an exiv2 project member you must rebase your repo copy (we use the name exiv2_upstream as an example below):

 $ git remote add exiv2_upstream https://github.com/Exiv2/exiv2.git
 $ git pull --rebase exiv2_upstream master
```

frli8848 commented:
@clanmills exiv2_upstream was just a random name I used to refer (reference) to the main https://github.com/Exiv2/exiv2.git repo.  This is needed to (I think) be able to update my copy (my fork) with the the changes made to main repo after I did the fork. The command:

$ git pull --rebase exiv2_upstream master

pulls in the changes and applies them to my fork/copy. But I'm very new to this so I can be wrong :-)

clanmills commented:
@frli8848 I don't understand git/GitHub enough to know how right you are.  I hope @Kicer86 will comment.  For sure, when this PR is committed, I'll test it by following your instructions.

D4N commented:
Fredrik Lingvall <notifications@github.com> writes:

> @D4N What is happening when you do the:
>
> $ git checkout -b my_feature_branch exiv2_upstream/master --no-track

This creates a new branch called my_feature_branch which is based on
exiv2_upstream/master (= it starts at the same commit) and then it
checks it out (i.e. switches to it). The --no-track is required so that
git does not automatically set exiv2_upstream/master as the default
branch to which you'll push (which will always fail, because no one can
push to exiv2_upstream/master, except via PRs).

There are many different ways how to achieve the same result, some of
which are more intuitive, like this one:
$ git checkout master
$ git pull exiv2_upstream/master
$ git branch my_feature_branch
$ git checkout my_feature_branch

However it can fail if you have screwed up your master branch and might
require a git reset --hard.

>
> The original CONTRIBUTING.md said that: " We prefer rebasing in top of master, instead of merging master into branches." Is this in conflict with the command above or is it a different thing?
>

That is a different thing. Albeit the quoted sentence is not entirely
correct.

>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/234#issuecomment-369904888

D4N commented:
Robin Mills <notifications@github.com> writes:

> @Kicer86 Can you review this, please.  This is an important document to establish the exact process for creating a PR.  There are other interesting/useful matters covered - however it's very important that the process to submit a PR is clarified and simple to follow.
>
> @frli8848 I think we should also document a procedure to enable folks to submit a patch.  For straightforward issues (such as your 3 one-line lens changes in February #227), a patch and test image should be sufficient.  We shouldn't require people to know and understand git/GitHub to make a one-line code change.
>

We can tell them that they should send a patch that was created via git
format-patch via email to one of the maintainers (my email address is in
the git commits, but I don't want it in plain sight on a webserver for
every spambot to find). Then I can apply & push it with relatively
little effort.

> @frli8848 Can you clarify the following please?  What is exiv2_upstream?
> ```
> After the PR has been reviewed by an exiv2 project member you must rebase your repo copy (we use the name exiv2_upstream as an example below):
>
>  $ git remote add exiv2_upstream https://github.com/Exiv2/exiv2.git
>  $ git pull --rebase exiv2_upstream master
> ``` 
>
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/234#issuecomment-369965930

D4N commented:
>  @clanmills exiv2_upstream was just a random name I used to refer (reference) to the main https://github.com/Exiv2/exiv2.git repo.  This is needed to (I think) be able to update my copy (my fork) with the the changes made to main repo after I did the fork. The command:
>
> $ git pull --rebase exiv2_upstream master
>
> pulls in the changes and applies them to my fork/copy. But I'm very new to this so I can be wrong :-)

You are correct, it updates your fork. There is a little more magic behind the `--rebase` flag. Consider the following example master branch with the commits:
```
A -> B -> C
```
where C is the most recent one. You start working on a pull request and create the additional commits D and E:
```
A -> B -> C -> D -> E
```
(that is now how your master branch looks like, not exiv2_upstream/master). No lets assume that I push commits F, G and H to exiv2_upstream/master resulting in this:
```
           /-> D -> E (your master)
A -> B -> C -> F -> G -> H (exiv2_upstream/master)
```
GitHub will now tell you in your PR that: "This branch is out-of-date with the base branch" and you can merge them via the button update branch. This creates a merge commit I (this is a commit that merges multiple branches):
```
           /-> D -> E ----/> I   (your master)
A -> B -> C -> F -> G -> H (exiv2_upstream/master)
```
And then we could merge the PR, resulting in:
```
           /-> D -> E ----/> I
A -> B -> C -> F -> G -> H ---\-> J
```
where J is one of the github commits of the form "Merge PR #SoAndSo".

In principle this approach is ok, but when PRs live long and you have multiple of these, you get a quite complicated web of commits. If you want to see an example, run `git log --oneline --graph` on the master branch and scroll down. Again, in principle no problem, but that makes backporting changes very hard, because you have to consider all the following & preceding merge commits.

Therefore we should try to do the following at the point where your master branch is out of date of the upstream master branch:
```
           /-> D -> E (your master)
A -> B -> C -> F -> G -> H (exiv2_upstream/master)
```
Now, don't merge the two branches, but you do the `git pull --rebase exiv2_upstream/master`. Git now takes your commits D and E and applies them directly onto the commit H, resulting in this:
```
                          /-> D' -> E' (your master)
A -> B -> C -> F -> G -> H (exiv2_upstream/master)
```
The downside of this is, that D and E no longer exist, git renamed them to D' & E' since their parent commit changed from C to H. That is also the reason, why you have to `push --force` afterwards (you have to force git to forget your commits D and E). As you can imagine, this can be now merged very easily:
```
                          /-> D' -> E' 
A -> B -> C -> F -> G -> H ---------\-> I
```
This has the big advantage, that there is only 1 strand of commits that can thus be very easily backported. Also Pull requests are now very easily visible in the git log. And you can in principle repeat the `git pull --rebase` as often as you like without creating an increasingly complicated history.

I hope that made sense and the ascii art is readable. I think I'll to write this whole thing down in a better form, so that is documented correctly.

D4N commented:
Robin Mills <notifications@github.com> writes:

> clanmills commented on this pull request.
>
> This is a beautiful job.  Thank You very much, Fredrik.  The TOC gives the document Structure and Style.  Thank You.
>
> Comments:
> 1) We only require the -fsanitize=address stuff as a check before submitting the PR.  It's not always needed.  What you've said is correct.  I'll leave you to decide if you should spell out the steps clone/build/develop/test/commit and only put the -fsanitize=address into test.
>

Unfortunately ASAN & UBSAN currently break the testsuite, so we probably
shouldn't suggest using them for running all the tests. (It's on my todo
list to fix this)

> 2) Should lines 19+20 be a single line `... such as, Wikis ...` And 72+73 `for non-programmers`.  And 137+138 `... bug report then create ...`
>
> 3) Can we put the stuff about patches into  Note: below the General Workflow
>
> 4) Capitalisation error on `7. And if you are asked...`
>
> 5) I don't understand git/github.  I think there's something missing concerning updating your fork if you want to make future submissions.  I believe your fork has to be updated.
>

You can keep your fork's master branch up to date with
exiv2_upstream/master by running:
$ git checkout master
$ git pull exiv2_upstream master

However, you don't have to do that. You can also work only in new
branches that you base of exiv2's master branch via:
$ git checkout -b feature_xyz exiv2_upstream/master --no-track

Once you are done with the PR, you can delete the branch and start a new
branch in which you can work. Then you don't even have to bother about
your fork's master branch. Plus you can easily switch between multiple PRs.

> 6) Responding to requests for changes to the PR.  If you've deleted your branch, you might have "cut yourself off" from making changes.  I'm not sure - something to have clarified.
>

That is not an easy thing to do, you have to delete your branch locally
and on GitHub. And even then it will be possible to recover your deleted
branch (as git never deletes commits immediately but waits 90 days),
albeit with more work involved using git reflog.

> Thank You very much for this great work.
>
>
>
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/234#pullrequestreview-100720863

clanmills commented:
@D4N  I'm happy to volunteer for the task of fixing the code to pass the test suite, once we have this CONTRIBUTING.md documentation finished and I am (finally) able to submit code to Exiv2.  I was aware that the test suite was broken and wondered if that was only on the Mac/clang.

So @frli8848, perhaps we should omit the `-DCMAKE=-fsanitize` for now and I'll add it later when it is working.

frli8848 commented:
An idea is to have a more detailed description of git process and the -fsanitize stuff in an appendix. There, we could also motivate why this is the preferred method to create PRs since it can be done in several different ways. 

Also, I have access to Apple hardware and I can probably test a more resent clang on my Gentoo box too (normally I run gcc 7.3.0 for full Meltdown/Spectre support).

clanmills commented:
I'll set up a user account for you on the buildserver (which is a Mac Mini).  You can ssh to the machine: exiv2.dyndns.org The machine sits on my desk!

Let me know your preferred username and I will set this up and send you a one-off/secure email with your password.  You may change the password when you access the machine.

The buildserver also has various Linux and MacOS-X VMs and a Windows 10 with all the MSVC compilers.  I'm the team's Windows expert!  You don't need to get involved with Windows unless you wish to do so.

frli8848 commented:
@clanmills you can set the username fl. I will update the draft for CONTRIBUTING.md file during next week. I will add two sections in the appendix, one for a detailed explantion and motivation for the PR process and one that describes the unit/CTests is detail. Perhaps @D4N can fill in/review those details in the appendix?

clanmills commented:
Sure.  I'll set that up this evening and email you the details.

I think this document is very important.  This PR is worth the effort to ensure that folks can easily contribute code to Exiv2 without being experienced users of git.

frli8848 commented:
BTW: Is there a way/tool to render GitHub's flavor of the .md files to html on Linux? Right now I have to push it to GitHub to see how it is rendered.

clanmills commented:
I don't know/understand git/GitHub to say what's possible.  You could try to raise a support ticket with them.

I think the git/GitHub philosophy is to check-in code **_on your fork_** frequently (every few minutes) and to push to 'master' occasionally (when you have a stable code-set).  When Luis submits a PR, it often has 20 steps.  So, push to your fork and look at it on the GitHub web-site.

Good News:  you're only working on one file at the moment, and it's documentation.  More complex changes involving code, build changes and updates to the test harness are more challenging.  So enjoy the experience because you're learning and growing.

Very importantly, you're doing something that will help many people.  I've been effectively unable to submit to Exiv2 since September 2017 because this document had not been written.

piponazo commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,121 @@
+
+1. Fork the Exiv2 git repository to your GitHub account by pressing the _Fork_ button at: [https://github.com/Exiv2/exiv2](https://github.com/Exiv2/exiv2) (more details [here](https://guides.github.com/activities/forking/)).
+
+1. Then start a terminal (or use your favorite git GUI app) and clone your copy of the Exiv2 repo:
```
[minor] This `1.` is duplicated. Please, update it to `2.` and the rest of the enumerated items

piponazo commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,121 @@
+
+2. Configure the project and build it (CMake is our the preferred configuration system, but not the only one [see [README.md](README.md)]):
+
+        $ cd exiv2
```
Could you fix the indentation here, please? And it would be also good to use a block of code (delimited with three back-ticks)

piponazo reviewed (changes requested):
Thanks a lot for this contribution @frli8848 ! I just requested few minor changes. We will approve the PR as soon as some of these minor issues are fixed.

piponazo commented:
I made my review before reading the comments in the PR. Just one clarification about the "git instructions". I said that I think this document might not be the best place to document "git instructions" because I think it could be important to keep the scope of the document.

However, if git usage is a pain for some of the contributors, I will yield in that matter for the sake of the git newcomers 😄

clanmills commented:
Wonderful.  It's not just newcomers!  The old Scottish guy needs a little help with the git monster.

I dont see anything in the current version of the document about creating a branch on your fork.  Is this needed or not?

D4N commented:
I would recommend doing that, unless you want to `pull --force` or start resetting. I think I wrote that further above.

How about we put this document in an etherpad and continue writing there? GitHub is not the best tool to work on such a document.

piponazo commented:
It is not obligatory but it is a recommended workflow.

This is how I normally work with git:

```
# Clone my fork
git clone https://github.com/piponazo/exiv2.git

# Add upstream to my list of remotes (https://help.github.com/articles/configuring-a-remote-for-a-fork/)
git remote add upstream https://github.com/Exiv2/exiv2.git

# This is just to assure that I have the two remotes configured correctly
$ git remote -v
origin	git@github.com:piponazo/exiv2.git (fetch)
origin	git@github.com:piponazo/exiv2.git (push)
upstream	git@github.com:Exiv2/exiv2.git (fetch)
upstream	git@github.com:Exiv2/exiv2.git (push)

# Create a branch where to work locally
git checkout -b myWonderfulBranch

# Do some work here
...
git commit -m "Commit 1"
...
git commit -m "Commit 2"

# Once I am happy with my work, create Pull Request (Using github, I do not use the command line for this)

# Let's say that my PR has been living for few days or weeks, reviewers requested few changes and people has merged other branches into upstream/master. Before resuming the work in my branch, I want to get the updates from upstream/master. https://help.github.com/articles/syncing-a-fork/

# First I fetch stuff from upstream (it does not matter in which branch I am currently on my local copy of the repository)
git fetch upstream

# I go to my master branch (the one corresponding to my fork)
git checkout master

# Now I merge the changes in upstream/master into my master branch
git merge upstream/master

# At this point I can push the changes to my fork (just to be sure that I do not need to do this again)
git push

# Now I go back to my branch
git checkout myWonderfulBranch

# And I rabase in top of master
git rebase master

# When we do a rebase, the commit history is rewritten and the next time you try to push your branch to your fork repository, you will need to use the --force option
git push --force
```

clanmills commented:
This looks like a great explanation and work-flow.  Could this be either added (as an appendix) or linked into CONTRIBUTING.md as "Recommended Git Workflow and Commands".

D4N commented:
As promised earlier, I have written the git guidelines. They are currently [here](https://github.com/D4N/exiv2/blob/git_guidelines/git_guidelines.md).

Since this adds further material to the discussion, I would propose that we at first finish the contributing.md and then discus the git_guidelines (in another PR). I posted it here so that we have rough idea what that document could cover to avoid overlaps.

clanmills commented:
Thanks, Dan.  I look at all of this tomorrow.  I agree.  We should have a different PR for "git guidelines".

frli8848 commented:
@D4N and @piponazo thanks for the excellent git explanations! I'm updating the doc right now and will push the changes today or tomorrow.

D4N commented:
Thanks for your work Fredrik. I'll take a look at the changes (hopefully) later today.

Also, you included commits from your previous PR in this one and now commits are present twice and the log is a mess. We'll have to fix that before we can merge this, but let's do that once the file contents are fixed.

@piponazo I think the Mac build on Travis is chocking on the hardcoded version for libcurl in conanfile.py. Looks like it can't download it. Do you think we should just use the latest available libcurl?

frli8848 commented:
@D4N Hm, I switched to another machine and git then "forced me" do do some merge:s - I'm a bit confused on how this really works... The commit 9e580c6483cef5d10facb54189776f2ae975a40c should be the most resent one. Also, I had to revert one commit since the kramdown renderer and GitHub's .md renderer translates to hmtl differently (the links on the TOC didn't work on GitHub). It would ease the pain if there was a .md tool that generates the same html as GitHub do.

frli8848 commented:
@D4N I suspect (but I'm not sure) that I did not rebase the clone on my laptop (I pushed the previous PR from my home work station) can this cause that commits are seen twice?

D4N commented:
I just found grip, a github markdown preview tool:
https://github.com/joeyespo/grip. It appears to be using the github API,
so the preview should be identical to what appears on github.

Concerning the additional commits, it comes down to that you did a pull
--rebase at one point on one machine but not on the other. If you run
git log --oneline --graph
you'll see the commits from your lens PR twice, first with the hashes:
cf290d33, 34f2a6ac, 19055c7d
and second with the hashes:
fc8b45dd, 34a3be02, 21522702

The first set are your original commits. Then I pushed to master and
asked you to git pull --rebase origin/master
git then dropped the commits cf290d33, 34f2a6ac, 19055c7d and reaplied
them, creating fc8b45dd, 34a3be02, 21522702.

Then you pushed -f to frli8848:master and everything looked good. The
issue is that your second machine had still the old commits cf290d33,
34f2a6ac, 19055c7d that were gone from frli8848:master. Now you pulled
from frli8848:master on that machine and git had to merge your commits
because your branches diverged.

This could have been only avoided by performing a pull -f (which
overwrites your local history) or by working in separate branches for
each PR (albeit then you can still get stale commits that got pushed
away).

Don't worry about it for now. We'll remove the old commits before we
merge this.


Fredrik Lingvall <notifications@github.com> writes:

> @D4N Hm, I switched to another machine and git then "forced me" do do some merge:s - I'm a bit confused on how this really works... The commit 9e580c6483cef5d10facb54189776f2ae975a40c should be the most resent one. Also, I had to revert one commit since the kramdown renderer and GitHub's .md renderer translates to hmtl differently (the links on the TOC didn't work on GitHub). It would ease the pain if there was a .md tool that generates the same html as GitHub do.
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/234#issuecomment-371139810

frli8848 commented:
@D4N Thanks for the info! I will try the grip tool.

clanmills commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,189 @@
+		$ rm -rf build
+		$ mkdir build
+		$ cd build
+		$ cmake --build . --config Release
```
$ cd build
$ cmake ..
$ cmake --build . --config Release

Also, I think there's a benign tab/formatting error

frli8848 replied:
Yes, and I was not sure about this either. I normally configure using -D args to cmake, like:

$ cmake -DCMAKE_BUILD_TYPE=Release ..
$ make

I think I use that.

D4N commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,189 @@
+See the [GIT_GUIDELINES.md](GIT_GUIDELINES.md) file for a more detailed description on the git workflow. 
+
+Below we outline the recommended steps in the code contribution workflow. We use `your-username` to refer to your username on GitHub, `exiv2_upstream` is used when we 
+set the upstream remote repository for Exiv2 (we could have picked any name except `origin` and `master` [TODO verify this]), and we use the name `my_first_pr_branch` 
```
I think it is possible for some name duplicates in git, but I certainly wouldn't recommend doing that. Nevertheless, I would just remove the sentence in brackets, as there are also other branches & tags in exiv2 whose name you shouldn't use and listing them all will get a little too much.

frli8848 replied:
OK. I will remove the TODO. Perhaps I can word it a little bit different so that people are aware of possible name collisions.

D4N commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,189 @@
+2. Then start a terminal (or use your favorite git GUI app) and clone your fork of the Exiv2 repo:
+
+        $  git clone https://github.com:your-username/exiv2.git
+
```
I think the creation of the new branch should be done here already, otherwise people will start hacking on the master branch only to realize they should be working on another one.

frli8848 replied:
Moved it to step 3.

D4N commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,189 @@
-### General ###
+	First, fetch stuff from exiv2_upstream (it does not matter in which branch you are currently in on your local copy of the repository):
+
+        $ git fetch exiv2_upstream
```
This is a quite verbose equivalent of `git checkout my_first_pr_branch && git pull --rebase origin master && git push --force`. I'd rather keep it short as there are fewer things that can be done wrong.

frli8848 replied:
I would prefer to to take each step separately (for beginners) but I can add a note at the end with the short version.

D4N replied:
Sure, but my point was that the steps that you copied from Luis are 7 in total, whereas the short version are only three. The end result is more or less the same but with only three steps there is less that can go wrong.

They can surely be written out as individual steps, but I'd rather include the short and simple version.

D4N commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,189 @@
+
+        $ git push --force
+
+Important: If your PR lives for a long time, then try to avoid to press the button _Update branch_ in the Pull Request view - we prefer rebasing in top of master, 
```
This is a remainder from the original that is ambiguous, I'd reword it in the following way (and maybe it should be moved further to the start of section 11):

Important: If your PR lives for a long time, then don't press the button _Update branch_ in the Pull Request view, instead follow the steps described above, as that avoids additional merge commits.

frli8848 replied:
OK, I can move it to the start of the section.

D4N commented on CONTRIBUTING.md:
```diff
@@ -70,8 +194,24 @@ Guidelines to apply clang-format:
+
+TODO: Add stuff here like:
+
+- intro to the test system (what we test and why),
```
System tests (= running exiv2 and comparing the output to some provided value) are documented under `tests/writing_tests.md` and `tests/doc.md`.

Unit tests use the Google Test framework, their official [Primer](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md) describes it pretty well.

What do you think should go in here additionally (I am asking because I know the systems and thus don't really see what someone not acquainted with it doesn't)?

clanmills replied:
Here's what I wrote a few years ago (before we had unit tests and the new python tests).  http://dev.exiv2.org/projects/exiv2/wiki/How_do_I_run_the_test_suite_for_Exiv2

D4N replied:
Oh, I didn't know we had `make teste` and `make testv`. Ops.

frli8848 replied:
OK, I will add links to the <tt>tests/writing_tests.md</tt> and <tt>tests/doc.md</tt> docs and remove the Appendix. If there are more stuff that is not covered in those docs (like Robin's wiki page) then <tt>tests/writing_tests.md</tt> and <tt>tests/doc.md</tt> should be updated in another PR right?

D4N replied:
Robin's Wiki page should be probably included anyway, since it covers the old testsuite under `test` whereas the two files cover the new test suite under `tests`

frli8848 replied:
OK, I will add that link too.

D4N commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,189 @@
+
+Once you have a GitHub login:
+
+1. Fork the Exiv2 git repository to your GitHub account by pressing the _Fork_ button at: [https://github.com/Exiv2/exiv2](https://github.com/Exiv2/exiv2) 
```
Exiv2 team members should work either in their fork or create a branch in the main repository to allow collaboration when applicable.

frli8848 replied:
Dan, to work on a branch on the main repo don't  one need write (push) rights to that repo?

D4N replied:
Yes, you need to have push rights. But if you have them, it can be beneficial to work in a branch in the main repository, since others can push there too. In your fork only you can, which makes it awkward if others want to add commits to a PR too.

frli8848 replied:
I will write a paragraph that explains that there are two methods but that we first will treat the "fork" method since I think it is important not to move too fast for beginners. It will anyway take a while before one gets push rights to the main repo, right?

D4N replied:
I think a remark, that people with push rights to the main repo should create branches in the exiv2 repo (when applicable), is enough. That way we have it documented also for team members and one-time contributors or newcommers don't get too confused by that.

Yes, it usually takes a while before you get push access.

frli8848 commented on CONTRIBUTING.md:
OK, I will add a link to the coding_guidelines.md doc. BTW should the file names be in upper- or lower case.

D4N replied:
I just looked at all other documentation files in the exiv2 root and they are uppercase, so it probably should be all-caps too.

frli8848 replied:
OK, so the rule is that all .md files in the root dir is uppercase and other files, like the ones in the tests dir, is lowercase?

D4N replied:
There is no rule. I made the .md files in tests lowercase because that's what I usually do. I would make the md files in the root dir all-caps for consistency with the rest.

frli8848 commented:
I just push:ed a new version to my fork. @D4N the grip tool was really helpful!

clanmills commented:
This looks really good.  When it passes the checks, we should integrate this.

We're in danger of turning this into "a ducks egg" (sit on this forever).  We have to get the duck to look away (at some of the referenced files such as GIT_GUIDELINES.md, doc.md, writing_tests.md) and steal the duck's egg!  Incidentally there are a couple of 404's in the generated code.  https://github.com/Exiv2/exiv2/pulls and GIT_GUIDELINES.md).

We don't need a perfect egg.  We've got a good one, so let's ship it.  We can polish it more later.

There's been some very good team-work to get this done.  Thank you Fredrik for your determination and effort.  Great Job Everybody.

frli8848 commented:
Yes, the GIT_GUIDELINES.md file do not exist yet. The doc.md, writing_tests.md links works for me at:

https://github.com/frli8848/exiv2/blob/master/CONTRIBUTING.md

I also had some more though on the PR process and, as Dan suggested, I think it would be good if we could reduce the number of steps involved a bit, but not too much. And, each step should have an easy to understand description. Also, should one do a new rebase for each time one push:es to the fork? If so then we need to describe that too.

D4N commented:
As a rule of thumb: in the branch in which you are working on the PR, you should only do `git pull --rebase origin master`. So yes, if you pull multiple times, you have to rebase multiple times too.

Concerning integrating this: please not yet. We **must** clean up the git history before we merge this.

frli8848 commented:
@D4N Should I do something to clean up the git history?

D4N commented:
At first, create a backup copy of your exiv2 folder and don't push to github unless you are certain it worked. 

The simplest solution is the following: create a copy of the files that you modified (that would be the contributing.md), do a `git reset --hard origin/master` (this will reset **everything** to the state of origin/master, so stash any unsaved changes, they'll be **gone** after that!), overwrite the contributing.md with your backup, commit the changes and `git push --force`. Preferably, take a look at the git history before pushing that it is the history of origin/master + the update of contributing.md.

This will however drop all your intermediate commits. Since they should be squashed anyway, I don't think that's a problem. If you want to keep them, then you'll have to do a little bit more work: create a new branch based on origin/master: `git checkout -b contributing_md_update exiv2_upstream/master --no-track`, now you have to cherry pick each commit from your master branch into this one. Only cherry pick commits with meaningful changes, i.e. no merges, like this: `git cherry-pick $commit_hash`. Once you have applied all your commits, compare the git log of `master` and `contributing_md_update` and verify that all your commits are still there but that the history is linear. Then go back to master: `git checkout master` and reset it to contributing_md_update: `git reset --hard contributing_md_update`.

frli8848 commented:
Dan, I have worked on the master (and not a branch) and push:ed it to my fork on GitHub. Does a   
<pre>
git reset --hard origin/master
</pre> 
help then? I could perhaps just delete the my fork on GitHub, Fork it again, clone it, and then push my changes? Or is it possible to "reset" the fork so it is identical to the upstream repo?

D4N commented:
`git reset --hard origin/master` will set the current branch (if you are in the `master` branch, then that one) to the same state as `origin/master`. In your case, you should probably use `exiv2_upstream/master` since you afaik have set github.com/Exiv2/exiv2.git as  exiv2_upstream.

But just to stress this out, `git reset --hard` will drop all your commits! Be sure to have a backup. (You can recover everything with git pull --force, provided you didn't push --force immediately after the reset).

I'll take a look at the current state of the document asap, but cannot promise it will be today.

frli8848 commented:
I will try to clean it up and I'm a bit curious how the cherry-pick stuff works so I will try that too.

D4N commented:
`cherry-pick` takes a commit and applies it on top of another branch. It is essentially the same what you'll get if you take the diff from a commit, make a patch out of it and apply the patch in another branch, only in one command.

It is mainly used to extract commits from development branches into stable branches, but you can also use it for other neat tricks.

D4N commented:
I have seen that you have removed the whole part about the coding guidelines, could you please move that into a new file `CODING_GUIDELINES.md`?

Also, I would advocate to use the simpler version in step 2.11.

clanmills commented:
@frli8848 If you adopt the 2.11 commands, I suggest you remove the &&'s and express the commands on different lines, with a comment on each command.  Something like:
```
$ cd ~/gnu/github/clanmills/exiv2
$ git remote -v                            # verify exiv2_upstream
$ git fetch exiv2_upstream                 # get into sync with exiv2/exiv2
$ git checkout master                      # work on the master branch of clanmills/exiv2
$ git merge exiv2_upstream/master          # sync master with exiv2_upstream
$ git checkout -b getopt-changes           # create a new branch
... restore my saved files and test ...
$ git push                                 # changes on username/exiv2 b:getopt-changes
```
I prefer to express command-blocks with the $ command prompt to discourage folks from blindly cut'n'pasting, however it's your document for you to format as you believe best.
```
cd ~/gnu/github/clanmills/exiv2
git remote -v                            # verify exiv2_upstream
git fetch exiv2_upstream                 # get into sync with exiv2/exiv2
git checkout master                      # work on the master branch of clanmills/exiv2
git merge exiv2_upstream/master          # sync master with exiv2_upstream
git checkout -b getopt-changes           # create a new branch
... restore my saved files and test ...
git push                                 # changes on username/exiv2 b:getopt-changes
```

frli8848 commented:
@D4N @clanmills I will add the CODING_GUIDELINES.md file and look at step 2.11 (I also prefer to have a $ before each command). I also think it would be good idea to have refs to these docs in the main README.md file.

clanmills commented:
@frli8848 Just go ahead and add/edit the files.  It's open-source, and you are welcome to contribute anything you think is of value.  The team review process is how code and documents are approved.

As you know, I'm a novice with git/GitHub and reluctant to approve merge anything into master.  I'm much happier to leave that to @D4N and @piponazo as they are our "git men".  Both joined the project in Summer 2017 have been involved in writing recent versions of those documents (and have made excellent contributions to the C++ code, build machinery and test harness.

In the past I have written documents concerning those topics and you will find them listed on http://dev.exiv2.org/projects/exiv2/wiki

If you wish to talk informally to any of us, you're welcome to chat on Slack.

frli8848 committed:
Restructured the file: added TOC, new sections: lens data and bug reports, and removed coding guidelines.

frli8848 committed:
Added a CODING_GUIDELINES.md file (which previously was a part of the CONTRIBUTING.md file).

frli8848 committed:
Simplified the git rebase step in the CONTRIBUTING.md file.

frli8848 commented:
I have now updated the CONTRIBUTING.md file, added the CODING_GUIDELINES.md file, and (hopefully) cleaned the git history. I dropped the cherry-pick stuff and just took the last version.

piponazo commented:
For me the PR looks in a good state to be merged. 

As @D4N pointed out there are some issues in the CI jobs related with Conan. I have created a issue about it (#243) and I will try to fix the issues within the next days.

Since this PR is not adding any code but just documentation changes, I would merge it without waiting for the CI to be green.

frli8848 committed:
Section numbering fixes.

clanmills commented:
Looking better all the time, Fredrik.  Thanks for your work.

A couple of comments:
1) In CONTRIBUTING.md
Let's not say 'my_first_pr'.  Let's say 'my_new_feature'
I believe git **_common practice_** is to use '-' in branch names, so probably **_my-new-feature_**

2) In writing_tests.md
Skip the discussion about python2.  It's written in python3.  If somebody requests to python2, we can invite them to join the team and contribute python2 support.

D4N commented on CONTRIBUTING.md:
This should be  `$ git pull --rebase exiv2_upstream master`.

frli8848 committed:
Changed name of the example branch.

frli8848 committed:
Fixed a bug in the rebase step.

frli8848 commented:
@clanmills I have now changed the name of the branch in the example but I have not touched the tests/writing_tests.md file - I think Dan wrote that file?

@D4N I have update the rebase step.

frli8848 commented:
@D4N We now have
<pre>
-snip- 
	First, fetch stuff from exiv2_upstream (it does not matter in which branch you are currently in on your local copy of the repository):

        $ git fetch exiv2_upstream

	Then switch to the branch

        $ git checkout my-new-feature

	And rebase in top of master:

        $ git pull --rebase exiv2_upstream master
</pre>
Should we both fetch and rebase from exiv2_upstream master or is one of the commands redundant?

D4N commented:
`git pull --rebase` will automatically perform a fetch, so you can skip the `git fetch` part.

frli8848 committed:
Simplified the git rebase description again.

frli8848 commented:
I have removed the `$ git fetch exiv2_upstream` command and updated the description accordingly. Is there any inconsistencies or anything missing now?

D4N commented:
This looks better then ever Fredrik, thank you for your work on this!

I think section 2.3 should also mention, that you absolutely have to start the new branch from `exiv2_upstream/master`, otherwise you'll drag in old commits. That can be achieved by doing a `git checkout exiv2_upstream/master` before the branch creation (you'll get a warning about a detached head, but in this case we can ignore that). I would add that as an additional note.

Before we merge this, can you once again pull --rebase and then squash your commits together?

D4N commented:
ping @frli8848 
Could you rebase on top of master & squash your commits into one?
I'll post a patch with the changes concerning the branches, if you are currently short on time.

frli8848 committed:
Added a note about branching from exiv2_upstream/master

frli8848 commented:
@D4N I have updated the branching description now but I have not tested it. I think I should start another PR just to test the procedure. We can update the instructions again if something shows up then? Perhaps @clanmills can test it too so we have two test samples :-)

D4N commented on CONTRIBUTING.md:
```diff
@@ -54,9 +54,11 @@ Once you have a GitHub login:
 
-        $ git checkout -b my-new-feature
+        $ git checkout exiv2_upstream/master
+		$ git checkout -b my-new-feature
```
You got two tabs instead of spaces here. Otherwise very nice!

frli8848 replied:
Hm, the emacs md-mode preview did not show this but some C-c C-s p magic fixed it.

D4N commented:
This looks quite complete to me, any objections to merging this @piponazo @clanmills ?

frli8848 committed:
Fixed a tabs to spaces issue.

D4N commented:
If it's fine with you all, I'd squash & merge all the commits.

piponazo commented:
Green light! Go ahead :smile:

D4N commented:
Thanks for your work on this Gentlemen, especially @frli8848 for playing the role of the etherpad!

frli8848 commented:
I had another look at the file and we do the:
<pre>
$ git checkout exiv2_upstream/master
</pre>
before the exiv2_upstream is defined. Step 5 in Section 2 should be step 3 etc.

frli8848 commented:
Yes - I will try!

Another thing, I tried the `git checkout exiv2_upstream/master` method on another project on GitHub and got the error:
<pre>
fl@nisse ~/projects/GoldenCheetah $ git remote add gc_upstream https://github.com/GoldenCheetah/GoldenCheetah.git
fl@nisse ~/projects/GoldenCheetah $ git checkout gc_upstream/master
error: pathspec 'gc_upstream/master' did not match any file(s) known to git.
</pre>
I have note tested it on exiv2 yet.

D4N commented:
Sorry, missed this notification. I think you have probably not fetched the remote branches. Try a `git fetch` after adding the remote.

frli8848 commented:
OK I will try that. I think I fix `git checkout exiv2_upstream/master` "bug" using a PR so that I can test the whole procedure. OK @D4N ?

frli8848 commented:
@D4N I have finally started to look at this (sorry for the delay) but I'm struggling with updating my fork on GitHub. I get merge conflicts when I'm trying to rebase it to upstream master even though the CONTRIBUTING.md files on https://github.com/exiv2/exiv2.git and  https://github.com/frli8848/exiv2.git are identical!? It looks like some old commits are pulled in.

frli8848 commented:
Here is (one of) my attempts:
<pre>
fl@nisse ~/projects $ rm -fR exiv2
fl@nisse ~/projects $ git clone https://github.com/frli8848/exiv2.git
Cloning into 'exiv2'...
remote: Counting objects: 34550, done.
remote: Total 34550 (delta 0), reused 0 (delta 0), pack-reused 34550
Receiving objects: 100% (34550/34550), 83.63 MiB | 2.04 MiB/s, done.
Resolving deltas: 100% (25912/25912), done.
fl@nisse ~/projects $ cd exiv2
fl@nisse ~/projects/exiv2 $ git remote add exiv2_upstream https://github.com/Exiv2/exiv2.git
fl@nisse ~/projects/exiv2 $ git fetch exiv2_upstream master                                 
remote: Counting objects: 195, done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 195 (delta 146), reused 171 (delta 146), pack-reused 24
Receiving objects: 100% (195/195), 37.70 KiB | 476.00 KiB/s, done.
Resolving deltas: 100% (146/146), completed with 47 local objects.
From https://github.com/Exiv2/exiv2
 * branch              master     -> FETCH_HEAD
 * [new branch]        master     -> exiv2_upstream/master
fl@nisse ~/projects/exiv2 $ git pull --rebase exiv2_upstream master
From https://github.com/Exiv2/exiv2
 * branch              master     -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
Applying: Restructured the file: added TOC, new sections: lens data and bug reports, and removed coding guidelines.
Using index info to reconstruct a base tree...
M       CONTRIBUTING.md
.git/rebase-apply/patch:92: trailing whitespace.
   
.git/rebase-apply/patch:97: trailing whitespace.
as, Wikis etc., will gradually be moved from Redmine to GitHub. User requests on the Redmine forums and the Redmine issue tracking system will be serviced until Exiv2 0.27 is released. 
.git/rebase-apply/patch:98: trailing whitespace.
After 0.27 is released the Redmine site will be read-only and all feature requests, bug reports etc., must be performed on GitHub. It is, however, preferred to use GitHub for all new 
.git/rebase-apply/patch:103: trailing whitespace.
Here we describe the recommended code contribution workflow after switching to GitHub. Code contributions is performed by using *pull requests* (PR) on GitHub. 
.git/rebase-apply/patch:104: trailing whitespace.
For this to work one first needs to [create a user account on GitHub](https://help.github.com/articles/signing-up-for-a-new-github-account/) if one do not 
warning: squelched 19 whitespace errors
warning: 24 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging CONTRIBUTING.md
CONFLICT (content): Merge conflict in CONTRIBUTING.md
error: Failed to merge in the changes.
Patch failed at 0001 Restructured the file: added TOC, new sections: lens data and bug reports, and removed coding guidelines.
The copy of the patch that failed is found in: .git/rebase-apply/patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
</pre>

frli8848 commented:
Ah, right - forgot that we  squashed all my commits into one here. Just for reference, this:
<pre>
fl@nisse ~/projects $ rm -fR exiv2
fl@nisse ~/projects $ git clone https://github.com/frli8848/exiv2.git
fl@nisse ~/projects/exiv2 $ git remote add exiv2_upstream https://github.com/Exiv2/exiv2.git
fl@nisse ~/projects/exiv2 $ git fetch exiv2_upstream master
fl@nisse ~/projects/exiv2 $ git checkout master
fl@nisse ~/projects/exiv2 $ git reset --hard exiv2_upstream/master
HEAD is now at 4bcec183 Merge pull request #286 from D4N/yet_another_testsuite_update
fl@nisse ~/projects/exiv2 $ git push origin master --force
</pre> 
made my fork identical to https://github.com/Exiv2/exiv2.git

frli8848 commented:
@D4N I just created a PR where I have updated the order of the git commands and, now I also have tested it :-) I will use it as a test to see if more updates/correction are needed to the CONTRUBUTING.md file (BTW the GIT_GUIDLINES.md file is still missing).
